tests/ashley/machina_extensions/forum/test_forum_view.py

```diff
@@ -0,0 +1,411 @@
+    """ Displays a forum and its topics overridden from django machina  """
+
+    def setUp(self):
+        super().setUp()
+
+        self.forum = ForumFactory()
+        # create 3 topics with distinct subject, views_count and date
+        PostFactory(
+            topic=TopicFactory(forum=self.forum, views_count=9),
+            subject="TOPIC B the eldest with 9 views_count",
+        )
+        PostFactory(
+            topic=TopicFactory(forum=self.forum, views_count=6),
+            subject="TOPIC A created second with 6 views_count",
+        )
+
+        PostFactory(
+            topic=TopicFactory(forum=self.forum, views_count=12),
+            subject="TOPIC C the newest one with 12 views_count",
+        )
+        user = UserFactory()
+        assign_perm("can_read_forum", user, self.forum)
+        self.client.force_login(user)
+
+        # Setup
+        self.url_list_topic = f"/forum/forum/{self.forum.slug}-{self.forum.pk}/"
+
+    def assertContentBefore(self, response, text1, text2, failing_msg=None):
+        """
+        Testing utility asserting that text1 appears before text2 in response
+        content.
+        """
+        self.assertLess(
+            response.content.index(text1.encode()),
+            response.content.index(text2.encode()),
+            (failing_msg or "")
+            + "\nResponse:\n"
+            + response.content.decode(response.charset),
+        )
+
+    def test_browsing_works(self):
+        """ default url should return a 200 """
+        # Run
+        response = self.client.get(self.url_list_topic)
+        # Check
+        assert response.status_code == 200
+
+    def test_default_sorting(self):
+        """ by default view is sorted by last_post_on """
+        # Run
+        response = self.client.get(self.url_list_topic)
+        # Should have 4 columns sortable
+        self.assertContains(response, "sortable ", count=4)
+        # Should have no sorted column, it's the default view
+        self.assertContains(response, "sorted ", count=0)
+        # Should be orderedy by date by default, the latest post first
+        self.assertContentBefore(
+            response, "TOPIC C the newest", "TOPIC A created second"
+        )
+        self.assertContentBefore(
+            response, "TOPIC A created second", "TOPIC B the eldest"
+        )
+
+        # Check we have expected default header links for CTAs
+        self.assertEqual(
+            response.context["header"],
+            TopicSortHeaderFactory.build_column_order(
+                [
+                    {
+                        "url_primary": "?o=0",
+                    },
+                    {
+                        "url_primary": "?o=1",
+                    },
+                    {
+                        "url_primary": "?o=2",
+                    },
+                    {
+                        "url_primary": "?o=3",
+                    },
+                ],
+            ),
+        )
+
```

Comment:
From what I understand, you are testing the context injected into the template.
Why not testing the HTML sent in the HTTP response ?

I understand that it's easier to test this way, but I'm worried because it implies that we consider that the templates can't contain errors.

(My comment applies to all tests using the response context)

Reply:
I've build it this way because testing links in the HTML is heavy and doesn't make me validate the building process. I've made a test that validates that what I see in my template corresponds to what I have in my links which certify these tests.